nixos/modules/services/networking/sozu.nix

```diff
@@ -0,0 +1,303 @@
+          };
+
+          pid_file_path = mkOption {
+            type = types.str;
+            default = "/run/sozu/sozu.pid";
+            description = ''
+              PID file is a file containing the PID of the master process of sozu.
+              It can be helpful to help systemd or any other service system to keep track
+              of the main process across upgrades. PID file is not created unless this option
+              is set or if SOZU_PID_FILE_PATH environment variable was defined at build time.
+            '';
+          };
+
+          tls_provider = mkOption {
+            type = types.enum [ "openssl" "rustls" ];
+            default = "rustls";
+            description = ''
+              Defines how the TLS protocol will be handled by sozu. Possible values
+              are "openssl" or "rustls". The "openssl" option will only work if sozu
+              was built with the "use-openssl" feature.
+            '';
+          };
+
+          front_timeout = mkOption {
+            type = types.int;
+            default = 60;
+            description = ''
+              Maximum time of inactivity for a front socket, in seconds.
+            '';
+          };
+
+          zombie_check_interval = mkOption {
+            type = types.int;
+            default = 1800;
+            description = ''
+              Duration between zombie checks, in seconds.
+              In case of bugs in sozu's event loop and protocol implementations,
+              some client sessions could be stuck, not receiving any more event,
+              and consuming resources. Sozu verifies regularly if there are such
+              zombie sessions, logs their state and removes them.
+            '';
+          };
+
+          activate_listeners = mkOption {
+            type = types.bool;
+            default = true;
+            example = false;
+            description = ''
+              By default, all listeners start a TCP listen socket on startup.
+              If set to false, this option will prevent them from listening. You can then add
+              the complete configuration, and send an ActivateListener message afterwards.
+            '';
+          };
+
+        };
+      };
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      description = ''
+        Extra configuration for <link xlink:href="https://github.com/sozu-proxy/sozu/blob/master/doc/configure.md#metrics">metrics</link>, <link xlink:href="https://github.com/sozu-proxy/sozu/blob/master/doc/configure.md#listeners">listeners</link>,
+        and <link xlink:href="https://github.com/sozu-proxy/sozu/blob/master/doc/configure.md#applications">applications</link>.
+      '';
+    };
+
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = with pkgs; [
+      sozu
+    ];
+
+    users.users.sozu = {
+      description = "Sozu Daemon User";
+      isSystemUser = true;
+    };
+
+    systemd.services.sozu = {
```

Comment:
Given the advertised reload ability with zero downtime that `sozu` boasts on its github page this unit seems tragic... **any** little change will cause this service to fully restart.

Reply:
Oh, I'm honestly not sure what I could improve here - I just copied & adjusted the unit provided by upstream.

Reply:
Have you validated that the `sozuctl ... reload` command works? I can't seem to have any success with it.

Here is my machine configuration:
```nix
  services.httpd.enable = true;
  services.httpd.adminAddr = "webmaster@example.org";

  services.httpd.virtualHosts.site0 = {
    listen = [ { ip = "0.0.0.0"; port = 8080; } ];
    documentRoot = pkgs.writeTextDir "index.html" "<html>this is site #0</html>";
  };

  services.httpd.virtualHosts.site1 = {
    listen = [ { ip = "0.0.0.0"; port = 8081; } ];
    documentRoot = pkgs.writeTextDir "index.html" "<html>this is site #1</html>";
  };

  services.httpd.virtualHosts.site2 = {
    listen = [ { ip = "0.0.0.0"; port = 8082; } ];
    documentRoot = pkgs.writeTextDir "index.html" "<html>this is site #2</html>";
  };

  services.httpd.virtualHosts.site3 = {
    listen = [ { ip = "0.0.0.0"; port = 8083; } ];
    documentRoot = pkgs.writeTextDir "index.html" "<html>this is site #3</html>";
  };

  services.sozu.enable = true;
  services.sozu.settings.applications."applications.localhost" = {
    protocol = "http";
    frontends = [ { address = "0.0.0.0:80"; hostname = "localhost"; } ];
    backends = [
      { address = "127.0.0.1:8080"; }
      # toggle a few of these and hit `curl localhost`
      # { address = "127.0.0.1:8081"; }
      # { address = "127.0.0.1:8082"; }
      { address = "127.0.0.1:8083"; }
    ];
  };
```

Additionally, I applied this patch on top of your PR:
```patch
diff --git a/nixos/modules/services/networking/sozu.nix b/nixos/modules/services/networking/sozu.nix
index 75be7642644..9589677eea0 100644
--- a/nixos/modules/services/networking/sozu.nix
+++ b/nixos/modules/services/networking/sozu.nix
@@ -31,6 +31,7 @@ in
         Documentation <link xlink:href="https://github.com/sozu-proxy/sozu/blob/master/doc/configure.md">here</link>.
       '';
       type = types.submodule {
+        freeformType = format.type;
 
         options = {
           saved_state = mkOption {
@@ -268,29 +269,22 @@ in
   };
 
   config = mkIf cfg.enable {
-    environment.systemPackages = with pkgs; [
-      sozu
-    ];
-
-    users.users.sozu = {
-      description = "Sozu Daemon User";
-      isSystemUser = true;
-    };
-
     systemd.services.sozu = {
       description = "Sozu - A HTTP reverse proxy, configurable at runtime, fast and safe, built in Rust.";
       after = [ "network.target" ];
       wants = [ "network.target" ];
       wantedBy = [ "multi-user.target" ];
 
+      reloadIfChanged = true;
+
       serviceConfig = {
+        DynamicUser = true;
         PIDFile = "/run/sozu/sozu.pid";
         ExecStart = "${cfg.package}/bin/sozu start --config ${configFile}";
         ExecReload = "${cfg.package}/bin/sozuctl --config ${configFile} reload";
         Restart = "on-failure";
-        User = "sozu";
         RuntimeDirectory = "sozu";
-        AmbientCapabilities = "CAP_NET_BIND_SERVICE";
+        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
       };
 
     };
```

**NOTE:** the only relevant parts of this patch are `systemd.services.sozu.reloadIfChanged = true;` and the addition of `freeformType = format.type;` to make my testing easier.

When I toggle some of the backends on and off, then `reload`, I still hit all 4 backends. Any ideas? I've never used `sozu` before, so maybe I'm missing something fundamental...

Reply:
I realized this is likely the expected behavior because you're not passing a static path to the configuration file. Actually the same thing exists with the `httpd` module currently... I should fix that.

Reply:
I fixed this issue in #122658. You should do the same thing here. I think we can get zero downtime updates with this software on NixOS.